panc/src/main/scripts/panlint/panlint.py

```diff
@@ -0,0 +1,363 @@
+
+    if strings:
+        for string in strings:
+            s, e = string.span()
+            debug_range(s, e, 'String Range')
+            string_ranges.append((s, e))
+
+    return string_ranges
+
+
+def filestats_table(problem_stats):
+    """Return a formatted table of problem counts per file"""
+    t = PrettyTable(['Filename', 'Problems'])
+    t.align['Filename'] = 'l'
+    t.align['Problems'] = 'r'
+    for filename, problem_count in problem_stats.iteritems():
+        if problem_count > 0:
+            t.add_row([filename, problem_count])
+    t.sortby = 'Filename'
+    return t
+
+
+def find_annotation_blocks(text):
+    """Find multi-line annotation blocks in a block of text"""
+    result = []
+    annotations = RE_ANNOTATION.finditer(text)
+    for annotation in annotations:
+        start_char, end_char = annotation.span()
+        start_line = text[:start_char].count('\n') + 1
+        end_line = start_line + text[start_char:end_char].count('\n')
+        for i in range(start_line, end_line + 1):
+            result.append(i)
+    return result
+
+
+def strip_trailing_comments(line, string_ranges):
+    for comment in RE_COMMENT.finditer(line):
+        # Does the candidate comment start inside a string?
+        # If so, it's not really a comment.
+        if not inside_string(comment.start(), comment.start()+1, string_ranges):
+            debug_range(comment.start(), comment.end(), 'Comment', False)
+            line = line[:comment.start()]
+    return line.rstrip()
+
+
+def lint_line(line, line_number, components_included, first_line=False):
+    debug_line(line, line_number)
+
+    messages = []
+    diagnoses = []
+    problem_count = 0
+
+    if first_line:
+        first_line = False
+        if not RE_FIRST_LINE.match(line):
+            messages.append('First non-comment line must be the template type and name')
+            diagnoses.append(diagnose(0, len(line)))
+            problem_count += 1
+
+    else:
+        string_ranges = get_string_ranges(line)
+        line = strip_trailing_comments(line, string_ranges)
+
+        for m in RE_COMPONENT_USE.finditer(line):
+            if m.group('name') not in components_included:
+                message = 'Component %s in use, but component config has not been included' % m.group('name')
+                diagnoses.append(diagnose(*m.span('name')))
+                messages.append(message)
+                problem_count += 1
+
+        for message, pattern in LINE_PATTERNS.iteritems():
+            m = pattern.search(line)
+            if m and m.group('error'):
+                start, end = m.span('error')
+                debug_range(start, end, 'Match', True)
+                if not inside_string(start, end, string_ranges):
+                    diagnoses.append(diagnose(start, end))
+                    messages.append(message)
+                    problem_count += 1
+
+        for _, check_method in getmembers(LineChecks(), predicate=ismethod):
```

Comment:
This underscore means "last thing the interpreter returned", it's not a donotcare variable.  Please rename this to

```python
for ignore, check_method in getmembers(LineChecks(), predicate=ismethod):
```